### DISCUSSION TITLE
How to play with promise ?

### DISCUSSION THREAD

**Discussion by ebocher** [caoccao/Javet#237] [Q&A]
I'd like to replace graaljs engine to javet  at this place https://github.com/orbisgis/demat/blob/main/src/main/java/org/orbisgis/demat/IOUtils.java#L98

I have tested different methods described in the javet doc  to play with promise , but without success.

Can you suggest a solution?

Best regards

**Comment by caoccao** [comment]
Please check out this file https://github.com/caoccao/Javet/blob/main/src/test/java/com/caoccao/javet/mock/MockFS.java for how to play with `Promise` in Javet. Also, you may visit this [doc](https://www.caoccao.com/Javet/reference/v8_values/v8_promise.html) for more detail. Please let me know if you have any questions.

**Comment by ebocher** [comment]
Thanks for the answer. Below is the pseudo code that works. I am not familiar with javascript so I' m not sure if it's the good way to do that. Can you give me your opinion. 

```
String svgFunction  = "var jsonFile;" +
                    "function svg() {\n" +
                    "var vegaspec = vegaLite.compile(JSON.parse(jsonFile)).spec;"+
                    "    var view = new vega.View(vega.parse(vegaspec))\n" +
                    "            .renderer('none')\n" +
                    "            .initialize()\n" +
                    "            .run()\n" +
                    "            ;\n" +
                    "    return view.toSVG();\n" +
                    "}";
            v8Runtime.getExecutor(svgFunction).executeVoid();

            File outputSVG = new File("/tmp/chart.svg");
            if(outputSVG.exists()){
                    outputSVG.delete();
            }
            try (V8ValueObject v8ValueObject = v8Runtime.createV8ValueObject()) {
                v8Runtime.getGlobalObject().set("jsonFile", v8ValueObject);
            }
            v8Runtime.getGlobalObject().set("jsonFile", here the json file);
            try (V8ValuePromise v8ValuePromise = v8Runtime.getExecutor(
                    "svg();").execute()) {
                if(v8ValuePromise.isFulfilled()){
                    V8Value result = v8ValuePromise.getResult();
                    try {
                        FileWriter fileWriter = new FileWriter(outputSVG);
                        fileWriter.write(result.toString());
                        fileWriter.close();
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
            v8Runtime.getGlobalObject().delete("jsonFile");           
```

**Reply by caoccao** [reply]
May I know which mode you use? The code looks good to me except you might want to call `await()` to pump the internal message queue.

**Comment by ebocher** [comment]
Thanks a lot. I'm using the default mode v8. If I use the node instance, it returns the following error message 
```
com.caoccao.javet.exceptions.JavetExecutionException: EvalError: Code generation from strings disallowed for this context
```

on 

```
V8ValuePromise v8ValuePromise = v8Runtime.getExecutor(
                    "svg();").execute()
```

**Reply by caoccao** [reply]
Please check this [doc](https://www.caoccao.com/Javet/reference/javadoc/com/caoccao/javet/interop/V8Runtime.html#allowEval-boolean-) out.

**Reply by ebocher** [reply]
Sorry not sure to understand what you mean ;-)

**Reply by caoccao** [reply]
`eval()` is disabled by default. That API can enable `eval()` so that your JS code will work.

**Reply by ebocher** [reply]
Ok. If I use node instance and set allowEval(true); the svg string is empty.

**Reply by caoccao** [reply]
I think you may try `await()`. Please review the `MockFS.java` in my first reply.

**Comment by ebocher** [comment]
I pushed two examples with V8  and nodejs modes here https://github.com/ebocher/demat/blob/main_next_iteration/src/test/java/org/orbisgis/demat/JavetTest.java

**Comment by ebocher** [comment]
moving await at the good place. It works !. Now the next challenge will be to save the chart in a png file. So I must find a way to rely on node canvas. Any suggestions are welcome ;-) 
By the way, javet rocks !

**Reply by caoccao** [reply]
I'm not familiar with that. But, `puppeteer` might be a good choice. And thank you for the compliment.
